Gems/GradientSignal/Code/Source/Editor/EditorImageGradientComponentMode.cpp

```diff
@@ -216,99 +373,134 @@ namespace GradientSignal
     void EditorImageGradientComponentMode::OnPaintBegin()
     {
         BeginUndoBatch();
+
+        // Get the spacing to map individual pixels to world space positions.
+        AZ::Vector2 imagePixelsPerMeter(0.0f);
+        ImageGradientRequestBus::EventResult(imagePixelsPerMeter, GetEntityId(), &ImageGradientRequestBus::Events::GetImagePixelsPerMeter);
+        if ((imagePixelsPerMeter.GetX() <= 0.0f) || (imagePixelsPerMeter.GetY() <= 0.0f))
+        {
+            return;
+        }
+
+        m_metersPerPixelX = 1.0f / imagePixelsPerMeter.GetX();
+        m_metersPerPixelY = 1.0f / imagePixelsPerMeter.GetY();
+
+        uint32_t imageWidth = 0;
+        ImageGradientRequestBus::EventResult(imageWidth, GetEntityId(), &ImageGradientRequestBus::Events::GetImageWidth);
+
+        uint32_t imageHeight = 0;
+        ImageGradientRequestBus::EventResult(imageHeight, GetEntityId(), &ImageGradientRequestBus::Events::GetImageHeight);
+
+        // Create the buffer for holding all the changes for a single continuous paint brush stroke.
+        // This buffer will get used during the stroke for ensuring we only affect each pixel once,
+        // and then after the stroke finishes we'll hand the buffer over to the undo system as an undo/redo buffer.
+        m_paintStrokeBuffer = AZStd::make_unique<ImageTileBuffer>(imageWidth, imageHeight, GetEntityId());
     }
 
     void EditorImageGradientComponentMode::OnPaintEnd()
     {
+        AZ_Assert(m_paintBrushUndoBuffer != nullptr, "Undo batch is expected to exist while painting");
+
+        // Expand the dirty region for this brush stroke by one pixel in each direction
+        // to account for any data affected by bilinear filtering as well.
+        m_paintStrokeDirtyRegion.Expand(AZ::Vector3(m_metersPerPixelX, m_metersPerPixelY, 0.0f));
+
+        // Expand the dirty region to encompass the full Z range since image gradients are 2D.
+        auto dirtyRegionMin = m_paintStrokeDirtyRegion.GetMin();
+        auto dirtyRegionMax = m_paintStrokeDirtyRegion.GetMax();
+        m_paintStrokeDirtyRegion.Set(
+            AZ::Vector3(dirtyRegionMin.GetX(), dirtyRegionMin.GetY(), AZStd::numeric_limits<float>::lowest()),
+            AZ::Vector3(dirtyRegionMax.GetX(), dirtyRegionMax.GetY(), AZStd::numeric_limits<float>::max()));
+
+        // Hand over ownership of the paint stroke buffer to the undo/redo buffer.
+        m_paintBrushUndoBuffer->SetUndoBufferAndDirtyArea(AZStd::move(m_paintStrokeBuffer), m_paintStrokeDirtyRegion);
+
         EndUndoBatch();
+
+        // Make sure we've cleared out our paint stroke data until the next paint stroke begins.
+        m_paintStrokeBuffer.reset();
+        m_metersPerPixelX = 0.0f;
+        m_metersPerPixelY = 0.0f;
+        m_paintStrokeDirtyRegion = AZ::Aabb::CreateNull();
```

Comment:
nit : might make sense to put these in their own separate struct that you could easily reset with `= {};`

Reply:
i'll structify it in my next PR.